Added use of insteon-hub streaming functionality

FreekingDean opened the pull request:
**Description:**


This adds the use of the Insteon streaming endpoint for physical device changes. It also defaults the system to not poll Insteon cloud to check device changes.

Since that is different than most other home automation toggleable status checking systems I would suggest that we make sure users are very informed as to why we decided to do this, and how they can change it if they would like.

I'd also like to not I have a ticket open with Insteon and will follow up with them on this approach. I would love feedback on this.  

**Related issue (if applicable):** fixes #3811 

**Pull request in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io) with documentation (if applicable):** home-assistant/home-assistant.github.io#1626

**Example entry for `configuration.yaml` (if applicable):**
```yaml
insteon_hub:
  username: YOUR_USERNAME
  password: YOUR_PASSWORD
  api_key: YOUR_API_KEY
  poll: SHOULD_POLL_CLOUD
```

**Checklist:**

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io)

If the code communicates with devices, web services, or third-party tools:
  - [x] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
  - [x] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [x] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [x] New dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [ ] New files were added to `.coveragerc`.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L16
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L51

FreekingDean committed:
Added use of insteon-hub streaming functionality, defaults to not poll cloud service

Also fixed doc & linting issues

FreekingDean committed:
Added callback for physical state change

balloob commented:
CC @wardcraigj @camrun91

balloob commented:
I don't think that we should even include the option to poll the cloud.

balloob commented on homeassistant/components/light/insteon_hub.py:
```diff
@@ -73,7 +92,10 @@ def turn_on(self, **kwargs):
         else:
             self._value = 100
             self.node.send_command('on')
+        self.update_ha_state()
```
Use `schedule_update_ha_state`

balloob replied:
and only call this if not polling. (which we should always do)

balloob commented on homeassistant/components/insteon_hub.py:
```diff
@@ -50,6 +49,17 @@ def setup(hass, config):
+        config
+    )
+    for insteon_house in INSTEON.houses:
+        threading.Thread(
```
How are these streams being stopped when home assistant stops?

FreekingDean replied:
There is a catch all exception handler in the stream method in `insteon-hub` that closes the stream.

balloob replied:
What happens locally when you are running Home Assistant with this version and you press control + c ? 

I am pretty sure that it will hang indefinitely as we are not cancelling the stream, hence the thread will keep running and keep Home Assistant running.

One way to work around it is to add `daemon=True`, that way the thread will not prevent Python from shutting down and thus get killed. The more correct way would be if we could call something like `cancel_stream` when the `EVENT_HOMEASSISTANT_STOP` is fired.

FreekingDean replied:
@balloob to do that would it be best to kill the thread in that case?

balloob commented on homeassistant/components/light/insteon_hub.py:
```diff
@@ -4,31 +4,39 @@
-            devs.append(InsteonToggleDevice(device))
-        if device.DeviceCategory == "Dimmable Lighting Control":
-            devs.append(InsteonToggleDevice(device))
+        if device.DeviceCategory in SUPPORTED_LIGHT_DEVICES:
```
```python
if device.DeviceCategory not in SUPPORTED_LIGHT_DEVICES:
    continue

new_device = …
```

balloob commented on homeassistant/components/light/insteon_hub.py:
```diff
@@ -47,18 +55,29 @@ def brightness(self):
 
+    def insteon_update(self, _):
+        """Callback for an update from the hosue streaming endpoint"""
+        self.update_ha_state()
```
This will not call `update`, you need to call schedule_update_ha_state(True)

FreekingDean committed:
Fix suggestion

Add period to end of comment/doc
Use schedule_update_ha_state over update_ha_state

FreekingDean commented on homeassistant/components/light/insteon_hub.py:
```diff
@@ -47,18 +55,29 @@ def brightness(self):
         try:
             self._value = resp['response']['level']
+            new_status = 'on' if self._value != 0 else 'off'
+            self.node.set_status(new_status)
```
This is setting a cached status of the device in the package. It does not assume a status based on the response as responses are very broad and can be interpreted different ways.

FreekingDean commented:
@balloob Is there a reason to not allow users to poll? That will allow them to have the most up to date info, if they use other applications (non hardware) to set Insteon devices this is the only way to know the devices current state.

balloob commented:
The polling code is what caused Insteon to request us to disable the integration.

balloob commented on homeassistant/components/light/insteon_hub.py:
```diff
@@ -4,31 +4,40 @@
     """An abstract Class for an Insteon node."""
 
-    def __init__(self, node):
+    def __init__(self, node, should_poll=False):
```
You don't use this override? Better to just remove it

balloob commented on homeassistant/components/insteon_hub.py:
```diff
@@ -5,25 +5,28 @@
         vol.Required(CONF_API_KEY): cv.string,
         vol.Required(CONF_PASSWORD): cv.string,
         vol.Required(CONF_USERNAME): cv.string,
+        vol.Optional(CONF_POLL): cv.boolean,
```
Please remove polling as an option. Insteon does not like it.

FreekingDean commented:
I do still want to wait to hear back from Insteon before doing too much more work on this.

I emailed them Monday, will try again next Monday.

balloob commented:
Any update on this ?

FreekingDean commented:
Still waiting to hear back. Sending another email over to them.

On Tue, Jan 17, 2017 at 3:21 AM Paulus Schoutsen <notifications@github.com>
wrote:

> Any update on this ?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/home-assistant/home-assistant/pull/4996#issuecomment-273049763>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABwE3bOxyS_Z9Oo3PQBXp9_byKZkW6oeks5rTHnwgaJpZM4LQa9u>
> .
>
-- 

Thanks,
--
Dean Galvin
Senior Software Engineer
Stitch Fix
https://www.stitchfix.com

homeassistant commented:
Hi @FreekingDean, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;home-assistant#4996).

Once you do that we will be able to review and accept this pull request.

Thanks!

homeassistant commented:
Hi @FreekingDean, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;home-assistant#4996).

Once you do that we will be able to review and accept this pull request.

Thanks!

balloob commented:
What's the plan when they don't answer?

As far as I can see you have addressed all their concerns with your PR.

FreekingDean commented:
Since Insteon is no longer (and hasn't for a while) been giving out API
keys I think we may want to remove Insteon cloud from Home Assistant core
all together. I can quickly throw it into a separate repo with
instructions. But I think that saying this is supported is not the best
approach. If they don't get back I vote we remove Insteon cloud and promote
insteon-local to 'insteon'.

On Thu, Feb 2, 2017 at 1:22 AM Paulus Schoutsen <notifications@github.com>
wrote:

What's the plan when they don't answer?

As far as I can see you have addressed all their concerns with your PR.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/home-assistant/home-assistant/pull/4996#issuecomment-276878337>,
or mute the thread
<https://github.com/notifications/unsubscribe-auth/ABwE3Qf1XoZSCxYphOPEefwMPEwitdqOks5rYXYxgaJpZM4LQa9u>
.

-- 

Thanks,
--
Dean Galvin
Senior Software Engineer
Stitch Fix
https://www.stitchfix.com

SEJeff commented:
Another option is to simply go full bore on making #6104 bulleproof and not using the Insteon Hub at all.

balloob commented:
Let's move forward with removing this component/platform from Home Assistant. Closing this PR.
